### PR TITLE
Allow multi-curative in second preventive optimisation

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

Second preventive RAO does not handle CRACs with multiple curative instants

**What is the new behavior (if this is a feature change)?**
It is now possible to run 2nd preventive RAO. However, impossible to fetch temporary results (for example the flow of CNEC monitored in curative2 after application of curative1 CRAs only)

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by bqth29** [comment]
@pet-mit @phiedw do you think it is still relevant to keep `FlowResult.getFlow(FlowCnec flowCnec, Side side, Unit unit)` now that we defined `FlowResult.getFlow(FlowCnec flowCnec, Side side, Unit unit, Instant instant)`?

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImpl.java [hunk 1]
```diff
@@ -281,7 +281,7 @@ public double getRelativeMargin(Instant optimizedInstant, FlowCnec flowCnec, Uni
     public double getFlow(Instant optimizedInstant, FlowCnec flowCnec, Side side, Unit unit) {
         FlowResult flowResult = getFlowResult(optimizedInstant, flowCnec);
         if (Objects.nonNull(flowResult)) {
-            return flowResult.getFlow(flowCnec, side, unit);
+            return flowResult.getFlow(flowCnec, side, unit, optimizedInstant);
```
Comment (phiedw) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImpl.java:284]: I think the method getStateOptimizedBefore also needs rewriting for curative instants (looking at the current code, it looks like it should be broken for anything but the first curative instant)

**Review by phiedw — CHANGES_REQUESTED** [review]

**Review by phiedw — DISMISSED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/api/FlowResult.java [hunk 2]
```diff
@@ -29,6 +30,17 @@ public interface FlowResult {
...
+    /**
+     * It gives the flow on a {@link FlowCnec}, at a given {@link com.powsybl.openrao.data.cracapi.Instant} and in a given {@link Unit}.
+     *
+     * @param flowCnec: The branch to be studied.
+     * @param side: The side of the branch to be queried.
+     * @param unit: The unit in which the flow is queried. Only accepted values are MEGAWATT or AMPERE.
+     * @param instant: The optimization instant for which the flow is queried.
+     * @return The flow on the branch in the given unit.
+     */
+    double getFlow(FlowCnec flowCnec, Side side, Unit unit, Instant instant);
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/api/FlowResult.java:42]:
```suggestion
    /**
     * It gives the flow on a {@link FlowCnec}, at a given {@link Instant} and in a given {@link Unit}.
     *
     * @param flowCnec: The branch to be studied.
     * @param side: The side of the branch to be queried.
     * @param unit: The unit in which the flow is queried. Only accepted values are MEGAWATT or AMPERE.
     * @param instant: The optimization instant for which the flow is queried.
     * @return The flow on the branch in the given unit.
     */
    double getFlow(FlowCnec flowCnec, Side side, Unit unit, Instant optimizedInstant);
```

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java [hunk 3]
```diff
@@ -673,22 +663,28 @@ private SecondPreventiveRaoResult runSecondPreventiveRao(RaoInput raoInput,
...
+            postContingencyResults.forEach((state, optimizationResult) -> {
+                if (state.getInstant().equals(instant)) {
+                    appliedRemedialActions.addAppliedNetworkActions(state, optimizationResult.getActivatedNetworkActions());
+                }
+            })
+        );
     }
 
-    static void addAppliedRangeActionsPostContingency(Instant instant, AppliedRemedialActions appliedRemedialActions, Map<State, OptimizationResult> postContingencyResults) {
+    // TODO add tests
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java:677]: add tests

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java [hunk 4]
```diff
@@ -673,22 +663,28 @@ private SecondPreventiveRaoResult runSecondPreventiveRao(RaoInput raoInput,
         return new SecondPreventiveRaoResult(secondPreventiveResult, postPraSensitivityAnalysisOutput, remedialActionsExcluded, appliedArasAndCras);
     }
 
-    static void addAppliedNetworkActionsPostContingency(Instant instant, AppliedRemedialActions appliedRemedialActions, Map<State, OptimizationResult> postContingencyResults) {
-        postContingencyResults.forEach((state, optimizationResult) -> {
-            if (state.getInstant().equals(instant)) {
-                appliedRemedialActions.addAppliedNetworkActions(state, optimizationResult.getActivatedNetworkActions());
-            }
-        });
+    // TODO add tests
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java:666]: add tests

**Review by bqth29 — APPROVED** [review]

**Review by bqth29 — APPROVED** [review]

**Pull request merged**